pyphare/pyphare/pharein/simulation.py

```diff
@@ -211,38 +211,93 @@ def check_origin(dims, **kwargs):
+            assert level_key[0] == "L"
+        else:
+            level_key = "L"+str(level_key)
+
+        if isinstance(boxes, list) and all([isinstance(box, Box) for box in boxes]):
+          list_per_level[level_key] = boxes
+
+        elif isinstance(boxes, dict):
+            list_per_level[level_key] = []
+
+            if all([isinstance(val, Box) for key,val in boxes.items()]):
+                for box_id, box in boxes.items():
+                    list_per_level[level_key] += [box]
+
+            elif all([isinstance(val, (tuple,list)) for key,val in boxes.items()]):
+                for box_id, box_coords in boxes.items():
+                    box_coords = [list(box_coord) for box_coord in box_coords]
+                    list_per_level[level_key] += [Box(box_coords[0], box_coords[1])]
+
+    return list_per_level
+
+def check_refinement_boxes(dims, **kwargs):
+
     refinement_boxes = kwargs.get("refinement_boxes", None)
 
     if refinement_boxes is None:
         return refinement_boxes
 
     smallest_patch_size = kwargs.get("smallest_patch_size")
+    refinement_ratio = kwargs["refinement_ratio"]
+    nesting_buffer = kwargs["nesting_buffer"]
+
+    refinement_boxes = as_list_per_level(refinement_boxes)
+    domain_box = Box([0] * dims, kwargs["cells"])
 
-    for ilvl, boxes in refinement_boxes.items():
+    boxes_per_level = {0: [domain_box]}
 
-        assert ilvl[0] == "L"
+    for level_key, boxes in refinement_boxes.items():
+
+        ilvl = int(level_key[1:])
+        refined_level_number = ilvl+1
+        boxes_per_level[refined_level_number] = boxes
 
-        refined_level_number = int(ilvl[1:])+1
         if (kwargs["max_nbr_levels"]) <= refined_level_number:
             err_msg = "Error - refinement boxes to create level {} invalid with 'max_nbr_levels'= {}"
             raise ValueError(err_msg.format(refined_level_number, kwargs["max_nbr_levels"]))
 
-        boxes = list(boxes.values())
         if len(boxes) == 0:
             raise ValueError("Error - missing refinement boxes")
 
-        tmp_boxes = []
-        if isinstance(boxes[0], (tuple,list)):
-            for points in boxes:
-                tmp_boxes.append(Box(points[0], points[1]))
-            boxes = tmp_boxes
-
-        elif isinstance(boxes[0], Box):
-            for box in boxes:
-                for l in boxm.refine(box, kwargs["refinement_ratio"]).length():
-                    if l < smallest_patch_size:
-                        raise  ValueError("Invalid box incompatible with smallest_patch_size")
+        for box_idx, ref_box in enumerate(boxes):
+            for cmp_box in boxes[box_idx + 1:]:
+                if ref_box * cmp_box != None:
+                    raise ValueError(f"Error: Box({ref_box}) overlaps with Box({cmp_box})")
+
+        for box in boxes:
+            coarser_boxes = boxes_per_level[ilvl]
+            if ilvl > 0:
+                coarser_boxes = [boxm.refine(box, refinement_ratio) for box in boxes_per_level[ilvl]]
```

Comment:
I imagine this will fail for level 2+, "refine()" should be called per level, or do something like 

"
boxm.refine(box, refinement_ratio ** ilvl)
"

maybe?